### ISSUE TITLE
Support for Apple M1 (`darwin-aarch64`)

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#2666]
# Update from 2023/02/01:

Support for Apple Silicon has landed in the 22.1 release of GraalVM. We gradually ship more GraalVM components for `darwin-aarch64` machines in new releases.

The latest state can be tested with [the latest GraalVM dev build](https://github.com/graalvm/graalvm-ce-dev-builds/releases/latest). Please help testing and [report bugs](https://github.com/oracle/graal/issues/new/choose)!

| Component       | Status            |
|-----------------|-------------------|
| Compiler    | experimental      |
| Native Image    | experimental      |
| LLVM runtime    | experimental      |
| LLVM toolchain  | experimental      |
| JavaScript      | experimental      |
| Node.js         | experimental |
| Java on Truffle | experimental      |
| Python          | experimental |
| Ruby            | experimental      |
| R               | planned           |
| WebAssembly     | planned           |
| VisualVM        | experimental      |

---

### Update from 2022/03/22:

Apple M1 support for [@GraalVM](https://twitter.com/graalvm) Native Image, libgraal, JavaScript, and Espresso [has landed](https://github.com/oracle/graal/issues/2666#issuecomment-1074884020)! 🎉🚀
M1 support for Ruby has also landed now.

You can now grab a GraalVM dev build for `darwin-aarch64` at: https://github.com/graalvm/graalvm-ce-dev-builds/releases/latest

Please help testing and [report bugs](https://github.com/oracle/graal/issues/new/choose)!

---

### Original Issue

**Describe the solution you'd like.**
Apple is transitioning to their [own chips](https://www.apple.com/newsroom/2020/06/apple-announces-mac-transition-to-apple-silicon/). GraalVM is already available for Linux+aarch64, we'll need macOS+aarch64/arm64 at some point too.

**Describe who do you think will benefit the most.**
Any future macOS user.

**Describe alternatives you've considered.**
I can confirm that GraalVM CE 20.1.0 for macOS+amd64 runs with [Rosetta 2](https://developer.apple.com/documentation/apple_silicon/about_the_rosetta_translation_environment) on a Developer Transition Kit (DTK).

**Additional context.**
[Apple documentation on Apple Silicon](https://developer.apple.com/documentation/apple_silicon). You can apply for a DTK [here](https://developer.apple.com/programs/universal/).

**Express whether you'd like to help contributing this feature**
I'm happy to help test GraalVM on a DTK. I've tried (but failed) to compile a LabsJDK for macOS+arm64 as a starting point.

**Comment by eregon** [comment]
The first step for this is to build a JVMCI JDK on darwin-arm64.

@dougxc told me the plan is to do this as soon as support is added for the darwin-arm64 platform in the JDK:
https://bugs.openjdk.java.net/browse/JDK-8253795

**Comment by chrisseaton** [comment]
Note that macOS 11 isn't part of the issue any more Graal works fine on that.

**Comment by eregon** [comment]
Internal issue: GR-27422

**Comment by eregon** [comment]
FWIW https://github.com/openjdk/jdk/pull/2200 got merged, but that's for JDK 17, planned to be released on 2021/09/14 (https://openjdk.java.net/projects/jdk/17/).
JVMCI JDK builds are always based on a release.
It might be backported to older JDKs, but that doesn't seem clear yet.
cc @deepj

**Comment by mutasemhidmi** [comment]
is it possible to build an image from graalvm arm64 for Linux in order to build native images. Note that I tried to do this to build a native image for Quarkus framework but it didn't work. I used this graalvm https://github.com/graalvm/container/tree/master/community in order to build my image, but eventually after running sam cli to test the lambda function I got this error msg="Init failed" InvokeID= error="fork/exec /var/task/bootstrap: no such file or directory"

**Comment by fniephaus** [comment]
I don't think that's possible. You can run native images built for macOS/AMD64 on Apple Silicon via Rosetta 2.

**Comment by mutasemhidmi** [comment]
@fniephaus I tried but the build got stuck, and I got a warning that I am using amd64 and not arm64. All the information is here 
[https://github.com/quarkusio/quarkus/issues/16225](https://github.com/quarkusio/quarkus/issues/16225)

**Comment by fniephaus** [comment]
Are you sure you used an amd64 GraalVM for macOS to generate the native image? Can you try building it on an Intel-based Mac? Then, copy over the native image and give it a go. If that doesn't work, you may be out of luck as an early adopter of an M1 Mac. I don't think native image support for Apple Silicon has a high priority for the GraalVM team at this point and will probably take weeks if not months to implement.

**Comment by mutasemhidmi** [comment]
@fniephaus yes I am sure. However, I think my case is a little different because eventually, I want to build an image that works on Linux(arm64/aarch64) since I want to deploy it on AWS Lambda. Note that I have docker desktop already downloaded, so I want to use this image to build my quarkus application Lambda function. I hope I expressed myself clearly.

**Comment by fniephaus** [comment]
Ah ok, so essentially, you want to build native image for Linux/aarch64 on your M1-powered Mac. You should just be able to do this with a Linux/aarch64 Docker container on your M1. Then just copy over the native image.

**Comment by deepj** [comment]
Is there any concrete plan to support Apple ARM's chip in GraalVM in near future since the support landed in OpenJDK (https://bugs.openjdk.java.net/browse/JDK-8253795)?

**Comment by dougxc** [comment]
> Is there any concrete plan to support Apple ARM's chip in GraalVM in near future since the support landed in OpenJDK (https://bugs.openjdk.java.net/browse/JDK-8253795)?

Please see this comment above: https://github.com/oracle/graal/issues/2666#issuecomment-817834239

**Comment by eregon** [comment]
To restate it a bit differently, I think we need to wait the OpenJDK 17 release (2021/09/14, https://github.com/oracle/graal/issues/2666#issuecomment-817834239) to build a JDK with darwin-aarch64 support. Unless darwin-aarch64 support gets backported earlier than that on JDK 8 or 11 (I have no idea).

**Comment by oraluben** [comment]
I tried to build GraalVM on M1, after hard-coded some architecture checks in mx, the first main issue is the absence of some dependency libraries on darwin arm64, and this can't be resolved easily by externals. Is there any plan that Graal team will add those dependencies (before the full support)?

**Comment by kammoh** [comment]
OpenJDK 17 has been released for a while, with full support for macos/arm64. Any updates on GraalVM?

**Comment by eregon** [comment]
Now JDK 17 is released and has darwin-aarch64 support.
The next issue is that [Apple only provides mac minis with 16GB ram](https://www.apple.com/us/shop/buy-mac/mac-mini), and that's basically not enough to do proper CI.
And without proper CI it's not really feasible to do development and testing for this architecture.

**Comment by mutasemhidmi** [comment]
Yeah but I mean to build it on a mac M1 in order to deploy it on Lambda.  The same way it's done now in Windows laptops.

**Comment by dougxc** [comment]
As we get sufficient hardware in our CI system to test and develop on darwin-aarch64, this will become a priority.
Any help from the GraalVM community before then would of course be very welcome.

**Comment by gtn1024** [comment]
Hope to see arm-based GraalVM on mac..

**Comment by debu999** [comment]
Recently started building on Apple silicon and will get help if arm64 is supported. Please advise if any traction. All new Macs are now with Apple silicon.

**Comment by mutasemhidmi** [comment]
Is there a timeline for this?

**Comment by dougxc** [comment]
Nothing that we can firmly commit to but I would expect something in the first quarter of 2022.

**Comment by dougxc** [comment]
> Is there anything users with apple silicon chips can do to help here? I.e. can we help out with the testing or whatever is required?

Yes!

The steps to try are:
* Clone https://github.com/graalvm/labs-openjdk-17 and build it using the steps in the `README.md`.
* Set `JAVA_HOME` to the result of that build and then run `mx --env ce build` in the `vm/` directory of the graal repo.

**Comment by cap10morgan** [comment]
@dougxc First issue from the labs-openjdk-17 README's build instructions:

```
configure: error: unrecognized options: --enable-aot
configure exiting with result code 1
```

I removed that arg and am proceeding with the rest of the directions.

**Comment by cap10morgan** [comment]
I had to add `--java-home` arg to the `mx` command, but then it failed like this:

```
$ mx --env ce --java-home $JAVA_HOME build
WARNING: The component inclusion list ('--components' or '$COMPONENTS') includes an unknown component: 'svml'
WARNING: The component inclusion list ('--components' or '$COMPONENTS') includes an unknown component: 'svml'
Downloading LLVM_ORG_SRC from ['https://lafo.ssw.uni-linz.ac.at/pub/llvm-org/llvm-src-llvmorg-12.0.1-3-g6e0a5672bc-bgf11ed69a5a.tar.gz']
 134340969 bytes (100%)
WARNING: No platform-specific definition is available for project com.oracle.truffle.llvm.tests.va.native for your architecture (aarch64)
WARNING: No platform-specific definition is available for library VISUALVM_PLATFORM_SPECIFIC for your architecture (aarch64)
library VISUALVM_PLATFORM_SPECIFIC:
Library without "path" attribute must have a non-empty "urls" list attribute or "maven" attribute
```

**Comment by dougxc** [comment]
It's probably best to try smaller steps.

First try building the compiler and running its unit tests:
```
cd compiler/
mx build
mx unittest
```

Then try building libgraal and running its gate:
```
cd vm/
mx --env libgraal build
mx --env libgraal gate -t "LibGraal"
```

**Comment by mbjarland** [comment]
Ok I'll add my 2c. Attempted the build. 

Below a few notes on the process for anybody reproducing this

(full terminal session log cat be found [in the following gist](https://gist.github.com/mbjarland/abe6f39897d5b13b48984893ffcb423c) , note the inline comments with `#` characters):

* cloned the repo from git@github.com:graalvm/labs-openjdk-17.git
* removed the `--enable-aot=no` flag from the `sh configure --with-conf-name=labsjdk...` line from the README.md
* the `configure` now failed with a `(Your Boot JDK version must be one of: 16 17)` error, set my JAVA_HOME and PATH to `sdkman-17.0.1-macosx_aarch64-zulu` which incidentally is a apple silicon zulu java 17 installation. 
* `configure` then works
* `make` also works

The following still says x86 at this point: 
```
╭─mbjarland@kirin ~/projects/labs-openjdk-17  ‹master›
╰─➤ ./build/labsjdk/images/graal-builder-jdk/bin/jshell
|  Welcome to JShell -- Version 17.0.2
|  For an introduction type: /help intro

jshell> System.out.println(System.getProperty("os.arch"));
x86_64
```
not sure what to make of that, in the apple silicon zulu installation we get this:

```
╭─mbjarland@kirin ~/projects/apple-silicon/graal/vm  ‹master›
╰─➤ jshell
|  Welcome to JShell -- Version 17.0.1
|  For an introduction type: /help intro

jshell> System.out.println(System.getProperty("os.arch"));
aarch64

jshell>
```
continuing on: 

* didn't have and wasn't familiar with `mx`. Installed via homebrew which apparently resulted in an incompatible version.
* cloned the mx repo from `git@github.com:graalvm/mx.git` and ran `mx` directly from the repo root which gave me `mx version 5.316.12`
* cloned the graal repo from `git@github.com:oracle/graal.git`
* `cd graal/vm` and executed the `mx` version from the mx repo `../../mx/mx --env ce build`
* build seems to have gotten pretty far but in the end fails with this: 
```
Archiving NATIVE_IMAGE_GRAALVM_SUPPORT... [/Users/mbjarland/projects/apple-silicon/graal/substratevm/mxbuild/darwin-amd64/dists/native-image-graalvm-support.tar does not exist]
JvmFuncsFallbacksBuildTask svm-jvmfuncs-fallback-builder... [/Users/mbjarland/projects/apple-silicon/graal/substratevm/mxbuild/jdk17/svm-jvmfuncs-fallback-builder/src_gen/JvmFuncsFallbacks.c does not exist]
Could not find any unresolved JVM_* symbols in static JDK libraries
JvmFuncsFallbacksBuildTask svm-jvmfuncs-fallback-builder: Failed due to error: 1
 819200 bytes (3%)Compiling com.oracle.truffle.llvm.toolchain.config with javac-daemon(JDK 17)... [/Users/mbjarland/projects/apple-silicon/graal/sulong/mxbuild/jdk17/com.oracle.truffle.llvm.toolchain.config/bin/com/oracle/truffle/llvm/toolchain/config/LLVMConfig.class does not exist]
 234089 bytes (100%)
 427381 bytes (100%))
 770048 bytes (58%)Downloading sources JAVACPP_SHADOWED from ['https://lafo.ssw.uni-linz.ac.at/pub/graal-external-deps/native-image/javacpp-shadowed-1.5.2-sources.jar']
 326165 bytes (100%))
 18251776 bytes (85%)
 21389286 bytes (100%)
 127068150 bytes (100%)
JvmFuncsFallbacksBuildTask svm-jvmfuncs-fallback-builder failed
1 build tasks failed
```
full log can be found in the gist above.

**Comment by mbjarland** [comment]
@dougxc - added file listing of `build/labsjdk/images/static-libs/lib/` to the comment above.

**Comment by dougxc** [comment]
The python crash can be avoided by making the build serial: `mx build -n`

Ok, looks like the `static-libs` target works so the problem must be in the part of `graal-builder-image` that copies the `*.a` final to the final image.
You should be able to do that manually with:
```
mkdir -p build/labsjdk/images/graal-builder-jdk/lib/static/darwin-aarch64
cp  build/labsjdk/images/static-libs/lib/*.a build/labsjdk/images/graal-builder-jdk/lib/static/darwin-aarch64
```

**Comment by mbjarland** [comment]
testing the above, will update here in a few

**Comment by mbjarland** [comment]
ok after copying the static libs and also setting my java home to: 
```
```
I get quite a lot further, full log can be found [here](https://gist.github.com/mbjarland/f8c2a3c3417104599d5a52c6a901212e), but the log ends with: 

```
========================================================================================================================
GraalVM Native Image: Generating 'graalvm-native-ld'...
========================================================================================================================
[1/7] Initializing...Error: Building images for org.graalvm.nativeimage.Platform$DARWIN_AMD64 requires static JDK libraries.
Use the JDK from https://github.com/graalvm/labs-openjdk-11/releases
Missing libraries: java, nio, net, zip
(search path: /Users/mbjarland/projects/apple-silicon/graal/sdk/mxbuild/darwin-amd64/GRAALVM_7EF73E63D7_JAVA17_STAGE1/graalvm-7ef73e63d7-java17-22.0.0-dev/lib)
Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
------------------------------------------------------------------------------------------------------------------------
                              0.1s spent in 6 GCs | Peak RSS: 0.51GB | CPU load: ~342.35%
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /Users/mbjarland/projects/apple-silicon/graal/sdk/mxbuild/darwin-amd64/graalvm-native-ld.image/graalvm-native-ld.build_artifacts.txt
========================================================================================================================
Finished generating 'graalvm-native-ld' in 3.2s.
Error: Image build request failed with exit status 1
```

**Comment by dougxc** [comment]
What's the output of `find /Users/mbjarland/projects/apple-silicon/graal/sdk/mxbuild/darwin-amd64/GRAALVM_7EF73E63D7_JAVA17_STAGE1/graalvm-7ef73e63d7-java17-22.0.0-dev/lib -name '*.a'`
If the output includes the missing libs, @olpaw may have some ideas as what the problem may be.

**Comment by mbjarland** [comment]
```
╭─mbjarland@kirin ~/projects/apple-silicon/graal/vm  ‹master›
╰─➤ cd /Users/mbjarland/projects/apple-silicon/graal/sdk/mxbuild/darwin-amd64/GRAALVM_7EF73E63D7_JAVA17_STAGE1/graalvm-7ef73e63d7-java17-22.0.0-dev/lib
╭─mbjarland@kirin ~/projects/apple-silicon/graal/sdk/mxbuild/darwin-amd64/GRAALVM_7EF73E63D7_JAVA17_STAGE1/graalvm-7ef73e63d7-java17-22.0.0-dev/lib  ‹master›
╰─➤ find . -name '*.a'
./svm/clibraries/darwin-amd64/liblibchelper.a
./svm/clibraries/darwin-amd64/libjvm.a
./svm/clibraries/darwin-amd64/libdarwin.a
./svm/builder/clibraries/darwin-amd64/liblibchelper.a
./svm/builder/clibraries/darwin-amd64/libjvm.a
./svm/builder/clibraries/darwin-amd64/libdarwin.a
./static/darwin-aarch64/libnet.a
./static/darwin-aarch64/libinstrument.a
./static/darwin-aarch64/libjava.a
./static/darwin-aarch64/libfreetype.a
./static/darwin-aarch64/libmanagement_agent.a
./static/darwin-aarch64/libmanagement_ext.a
./static/darwin-aarch64/libattach.a
./static/darwin-aarch64/libj2pcsc.a
./static/darwin-aarch64/libosxapp.a
./static/darwin-aarch64/libjdwp.a
./static/darwin-aarch64/libfontmanager.a
./static/darwin-aarch64/libfdlibm.a
./static/darwin-aarch64/libjaas.a
./static/darwin-aarch64/libjli.a
./static/darwin-aarch64/libsyslookup.a
./static/darwin-aarch64/libjawt.a
./static/darwin-aarch64/libosxsecurity.a
./static/darwin-aarch64/libjsound.a
./static/darwin-aarch64/libjsig.a
./static/darwin-aarch64/libsaproc.a
./static/darwin-aarch64/libzip.a
./static/darwin-aarch64/libjavajpeg.a
./static/darwin-aarch64/libverify.a
./static/darwin-aarch64/libmanagement.a
./static/darwin-aarch64/libjimage.a
./static/darwin-aarch64/libawt_lwawt.a
./static/darwin-aarch64/libawt.a
./static/darwin-aarch64/libsplashscreen.a
./static/darwin-aarch64/liblcms.a
./static/darwin-aarch64/libextnet.a
./static/darwin-aarch64/libj2pkcs11.a
./static/darwin-aarch64/librmi.a
./static/darwin-aarch64/libjpackageapplauncher.a
./static/darwin-aarch64/libosx.a
./static/darwin-aarch64/libdt_socket.a
./static/darwin-aarch64/libosxui.a
./static/darwin-aarch64/libmlib_image.a
./static/darwin-aarch64/libprefs.a
./static/darwin-aarch64/libnio.a
./static/darwin-aarch64/libosxkrb5.a
./static/darwin-aarch64/libj2gss.a
╭─mbjarland@kirin ~/projects/apple-silicon/graal/sdk/mxbuild/darwin-amd64/GRAALVM_7EF73E63D7_JAVA17_STAGE1/graalvm-7ef73e63d7-java17-22.0.0-dev/lib  ‹master›
╰─➤
```

**Comment by dougxc** [comment]
I think the problem is [here](https://github.com/oracle/graal/blob/df2133fa08f60599ba97a687e5a3b5f63f071c63/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java#L297) based on your [earlier comment](https://github.com/oracle/graal/issues/2666#issuecomment-982581998) about `os.arch` reporting `x86_64`.
So the next thing is to figure out why that happens.

**Comment by mbjarland** [comment]
Trying to trace that one down seems to be a rabbit hole (at least to my very uninformed eyes). [This](https://github.com/graalvm/labs-openjdk-17/blob/13e2a145487df3194fe82ce4299312f5f3e13766/src/java.base/share/classes/jdk/internal/util/SystemProps.java#L77) seems relevant but then we immediately reach out to native code via the `Raw` class [here](https://github.com/graalvm/labs-openjdk-17/blob/13e2a145487df3194fe82ce4299312f5f3e13766/src/java.base/share/classes/jdk/internal/util/SystemProps.java#L183-L208)...which is pretty much where my skill set ends and black voodoo magic begins : ).

**Comment by cap10morgan** [comment]
When I run `System.out.println(System.getProperty("os.arch"));` in my built labs-openjdk-17 jshell I get `aarch64`. So I'll follow the steps from there and see if I get a different result.

**Comment by cap10morgan** [comment]
> I think the problem is [here](https://github.com/oracle/graal/blob/df2133fa08f60599ba97a687e5a3b5f63f071c63/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java#L297) based on your [earlier comment](https://github.com/oracle/graal/issues/2666#issuecomment-982581998) about `os.arch` reporting `x86_64`.
> So the next thing is to figure out why that happens.

@mbjarland The terminal where you built your labs-openjdk binaries isn't running under Rosetta 2 is it? You can check in Activity Monitor. Just a thought.

**Comment by mbjarland** [comment]
@cap10morgan good call...but no cigar it seems. I'm running iterm2 and checking activity monitor it reports the iterm2 process as "Kind" "Apple" ...assuming that should suffice?

**Comment by cap10morgan** [comment]
> @cap10morgan good call...but no cigar it seems. I'm running iterm2 and checking activity monitor it reports the iterm2 process as "Kind" "Apple" ...assuming that should suffice?

Yeah you should be good then. 🤔

The only other obvious difference I can see between my system and yours is that I used Homebrew's `openjdk` formula (which is currently version 17.0.1) as my bootstrap JDK when I built my labs-openjdk-17 binaries.

**Comment by cap10morgan** [comment]
libgraal results:

`mx --env libgraal --java-home $JAVA_HOME build -n` seemed to succeed.

`mx --env libgraal --java-home $JAVA_HOME gate -t "LibGraal"` did not succeed. It's output isn't so long, so I'll post it here:

```
WARNING: The component inclusion list ('--components' or '$COMPONENTS') includes an unknown component: 'svml'
WARNING: The component inclusion list ('--components' or '$COMPONENTS') includes an unknown component: 'svml'
gate: 30 Nov 2021 08:46:57(+00:00) BEGIN: Gate
gate: 30 Nov 2021 08:46:57(+00:00) BEGIN: LibGraal Compiler:GraalVM DaCapo-avrora
Downloading DACAPO from ['https://lafo.ssw.uni-linz.ac.at/pub/graal-external-deps/dacapo-9.12-bach-patched.jar']
 167107372 bytes (100%)
===== DaCapo 9.12 avrora starting =====
===== DaCapo 9.12 avrora PASSED in 2473 msec =====
Verifying content of /Users/wmorgan/dev/graal/vm/dacapo-work.M0itEN/hs_err_pid39070.log
Verifying content of /Users/wmorgan/dev/graal/vm/dacapo-work.M0itEN/hs_err_pid39070_libjvmci.log
Cleaning up scratch dir after gate task completion: /Users/wmorgan/dev/graal/vm/dacapo-work.M0itEN
gate: 30 Nov 2021 08:48:05(+01:08) END:   LibGraal Compiler:GraalVM DaCapo-avrora [0:01:08.102662] [disk (free/total): 1.3TB/1.8TB]
gate: 30 Nov 2021 08:48:05(+01:08) BEGIN: LibGraal Compiler:CTW
WARNING: Using incubator modules: jdk.incubator.vector, jdk.incubator.foreign
[Use -Dgraal.LogFile=<path> to redirect Graal log output to a file.]
CompileTheWorld : iteration 0
CompileTheWorld : Gathering compilations ...
Monitor starting
CompileTheWorld : Using 10 threads
CompileTheWorld : Starting compilations ...
[thread 59395 also had an error]#

[thread 60931 also had an error]# A fatal error has been detected by the Java Runtime Environment:

#
#  SIGSEGV (0xb) at pc=0x00000001680fc978, pid=39076, tid=59651
#
# JRE version: OpenJDK Runtime Environment GraalVM 22.0.0-dev (17.0.2) (build 17.0.2+0-jvmci-22.0-b02)
[thread 69891 also had an error]# Java VM: OpenJDK 64-Bit Server VM GraalVM 22.0.0-dev (17.0.2+0-jvmci-22.0-b02, mixed mode, sharing, tiered, jvmci, jvmci compiler, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)

# Problematic frame:
# [thread 62211 also had an error][thread 61443 also had an error][thread 69379 also had an error][thread 61699 also had an error]


[thread 68867 also had an error]

[thread 60163 also had an error]
C  [libjvmcicompiler.dylib+0xfc978]  LibGraalEntryPoints_compileMethod_21f8f52578424b633d4de12d1750b244fd89cf58+0x6b8
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/wmorgan/dev/graal/vm/hs_err_pid39076.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/oracle/graal/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

gate: 30 Nov 2021 08:48:12(+01:15) END:   LibGraal Compiler:CTW [0:00:07.307862] [disk (free/total): 1.3TB/1.8TB]
Traceback (most recent call last):
  File "/Users/wmorgan/dev/mx/mx_gate.py", line 422, in gate
    _run_gate(cleanArgs, args, tasks)
  File "/Users/wmorgan/dev/mx/mx_gate.py", line 651, in _run_gate
    runner(args, tasks)
  File "/Users/wmorgan/dev/graal/vm/mx.vm/mx_vm_gate.py", line 155, in gate_body
    ], extra_vm_arguments)
  File "/Users/wmorgan/dev/graal/compiler/mx.compiler/mx_compiler.py", line 322, in ctw
    run_vm(vmargs + mainClassAndArgs)
  File "/Users/wmorgan/dev/graal/compiler/mx.compiler/mx_compiler.py", line 1096, in run_vm
    return run_java(args, nonZeroIsFatal=nonZeroIsFatal, out=out, err=err, cwd=cwd, timeout=timeout)
  File "/Users/wmorgan/dev/graal/compiler/mx.compiler/mx_compiler.py", line 1058, in run_java
    return mx.run(cmd, nonZeroIsFatal=nonZeroIsFatal, out=u.out, err=u.err, timeout=timeout, cwd=cwd, env=env)
  File "/Users/wmorgan/dev/mx/mx.py", line 13205, in run
    abort(retcode)
  File "/Users/wmorgan/dev/mx/mx.py", line 4139, in abort
    raise SystemExit(error_code)
SystemExit: -6

The sequence of mx commands that were executed until the failure follows:


If the previous sequence is incomplete or some commands were executed programmatically use:

mx --env libgraal --java-home /Users/wmorgan/dev/labs-openjdk-17/build/labsjdk/images/graal-builder-jdk gate -t LibGraal

gate: 30 Nov 2021 08:48:12(+01:15) ABORT: Gate [0:01:15.414680] [disk (free/total): 1.3TB/1.8TB]
-6
```

**Comment by mbjarland** [comment]
@cap10morgan what does your `sh configure ...` command report at the end? 

Mine has this: 

```
Configuration summary:
* Name:           labsjdk
* Debug level:    release
* HS debug level: product
* JVM variants:   server
* JVM features:   server: 'cds compiler1 compiler2 dtrace epsilongc g1gc jfr jni-check jvmci jvmti management nmt parallelgc serialgc services shenandoahgc vm-structs zgc'
* OpenJDK target: OS: macosx, CPU architecture: x86, address length: 64
* Version string: 17.0.2+0-jvmci-22.0-b02 (17.0.2)
```

**Comment by dougxc** [comment]
@cap10morgan can you please attach `hs_err_pid39076.log` or make it available somewhere.

**Comment by cap10morgan** [comment]
> @cap10morgan what does your `sh configure ...` command report at the end?
> 
> Mine has this:
> 
> ```
> Configuration summary:
> * Name:           labsjdk
> * Debug level:    release
> * HS debug level: product
> * JVM variants:   server
> * JVM features:   server: 'cds compiler1 compiler2 dtrace epsilongc g1gc jfr jni-check jvmci jvmti management nmt parallelgc serialgc services shenandoahgc vm-structs zgc'
> * OpenJDK target: OS: macosx, CPU architecture: x86, address length: 64
> * Version string: 17.0.2+0-jvmci-22.0-b02 (17.0.2)
> ```
> 
> of special note on the second to last line:
> 
> ```
> * OpenJDK target: OS: macosx, CPU architecture: x86, address length: 64
> ```

@mbjarland I get this:

```
Configuration summary:
* Name:           labsjdk
* Debug level:    release
* HS debug level: product
* JVM variants:   server
* JVM features:   server: 'cds compiler1 compiler2 dtrace epsilongc g1gc jfr jni-check jvmci jvmti management nmt parallelgc serialgc services shenandoahgc vm-structs zgc'
* OpenJDK target: OS: macosx, CPU architecture: aarch64, address length: 64
* Version string: 17.0.2+0-jvmci-22.0-b02 (17.0.2)
```

**Comment by cap10morgan** [comment]
> @cap10morgan can you please attach `hs_err_pid39076.log` or make it available somewhere.

Yes: https://gist.github.com/cap10morgan/1f5918b75f9d5a52f591d3df35919fa6

**Comment by mbjarland** [comment]
ok apologies for the detour above, my `zsh` which is the shell I used to execute things was running under rosetta and intel. Needed to do a song and dance (uninstall homebrew, reinstall, reinstall all homebrew installed things as apple versions etc) to remedy that. Almost guaranteed to have caused the misbehavior above. 

Will re-try once I've fixed all that.

**Comment by lewurm** [comment]
FYI you can switch between architectures on the CLI like this:

```sh
$ arch -x86_64 bash
$ arch
i386
$ arch -arm64 bash
$ arch
arm64
```

I think there are four relevant scenarios:
1. Run in `x86_64` mode and use a `x86_64` boot JDK. This should work today imho as it just means everything runs under Rosetta 2. However, this is not what this issue is asking for.
2. Run in `x86_64` mode and use a `arm64` boot JDK. This is kind of cross-compilation. While theoretically possible, I don't recommend going down this path.
3. Run in `arm64` mode and use a `x86_64` boot JDK. Same as `2.`, don't do that.
4. Run in `arm64` mode and use a `arm64` boot JDK. This is what this issue is about: Run everything natively without Rosetta 2 involvement.

**Comment by mbjarland** [comment]
Link to contents of [/Users/mbjarland/projects/apple-silicon/graal/vm/hs_err_pid12942.log](https://gist.github.com/mbjarland/8d1170bec489521914fea37c321d6587) in the above execution.

**Comment by dougxc** [comment]
Thanks for the hs-err logs. At this point, it will require someone with access to an M1 Mac and working knowledge of libgraal to get further. There's no one in that position yet on the compiler team but hopefully that will be resolved early in 2022.

**Comment by mbjarland** [comment]
Perhaps something like this: 

https://www.scaleway.com/en/hello-m1/

might be useful?

**Comment by mbjarland** [comment]
Ok so I have an apple silicon m1 instance running in the cloud. Could somebody appropriate ping me via email at `mbjarland@gmail.com` and I can provide appropriate connection details. I can leave that instance up at least for a month and possibly more.

**Comment by mbjarland** [comment]
@dougxc any takers for the above? I would need the public ssh key of whoever wants to connect and perhaps just ping them with how things are setup, but that should be it. 

Also I realize that 8GB ram is low for CI (as @eregon mentioned above, seems even 16 might be too low), but perhaps enough to figure out any potential build issues? 

Scaleway only has the 8GB M1 instance type. There seems to be a few others (hetzner, macstadium) with 16G M1 mac mini instances which might or might not be enough for your CI needs.

**Comment by dougxc** [comment]
Thanks for setting up the scaleway instance. It may be interesting to someone in the community to use.
At Oracle, we have a couple of Mac Mini instances that were added to our internal pool of resources recently. There's a chance @lewurm may be able to spend time with them soon to progress this effort further.

**Comment by mbjarland** [comment]
Ok no worries, I'll retire that instance. Just a thought.

**Comment by teshull** [comment]
Thanks for looking into things. W.r.t.  LibGraal, internally we need to do some testing to help ensure native image is working properly on Darwin AArch64. #4077 addresses one issue, but there may be others.

For now it'd probably be best to test things with JarGraal

**Comment by lewurm** [comment]
> Compiler unit tests produced a lot of output ([full output here](https://gist.github.com/cap10morgan/d7b55f1165eb6466e2d6b2e1e0607c69)), but the final result was:
> 
> ```
> FAILURES!!!
> Tests run: 43214,  Failures: 25
> ```

The NFI related failures should be fixed by https://github.com/oracle/graal/commit/4d61ca57e601d74b92574b333a2cce40d9fa780f

**Comment by mbjarland** [comment]
For the little less initiated among us, what does the above mean for the overall progress of this issue?

**Comment by lewurm** [comment]
Status update:
1. The PR https://github.com/oracle/graal/pull/4173  fixes `mx --env libgraal gate` in `vm/`.  There are some issues with the PR in its current state (e.g. it won't handle floating points correctly), but that should be sorted out soon.
2. In order to run Truffle languages via SubstrateVM, we need to add support for `MAP_JIT` [1] to enable just-in-time compilation. From what I've seen so far this doesn't look too hard as SubstrateVM doesn't do any code patching (at least if `WriteableCodeCache` isn't used, which isn't supported on AArch64 today), and therefore maintaing the W/X state via `pthread_jit_write_protect_np()` should be straight forward.
3. JDK17 / `labs-openjdk-17` supports darwin-aarch64 today, for JDK11 there is a pending backport https://github.com/openjdk/jdk11u-dev/pull/715, so `labs-openjdk-11` should be ready for darwin-aarch64 soon as well.
4. We are starting to hook up M1 machines to our build infrastructure in order to gradually produce images (starting with `labs-openjdk-17`).


[1] https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon

**Comment by ebytez** [comment]
Thanks for the update Bernhard (and on Slack too)!

> Status update:
> 
> 1. The PR [[GR-36135] darwin-aarch64: respect native ABI for incoming args too #4173](https://github.com/oracle/graal/pull/4173)  fixes `mx --env libgraal gate` in `vm/`.  There are some issues with the PR in its current state (e.g. it won't handle floating points correctly), but that should be sorted out soon.
> 2. In order to run Truffle languages via SubstrateVM, we need to add support for `MAP_JIT` [1] to enable just-in-time compilation. From what I've seen so far this doesn't look too hard as SubstrateVM doesn't do any code patching (at least if `WriteableCodeCache` isn't used, which isn't supported on AArch64 today), and therefore maintaing the W/X state via `pthread_jit_write_protect_np()` should be straight forward.
> 3. JDK17 / `labs-openjdk-17` supports darwin-aarch64 today, for JDK11 there is a pending backport [8253795: Implementation of JEP 391: macOS/AArch64 Port openjdk/jdk11u-dev#715](https://github.com/openjdk/jdk11u-dev/pull/715), so `labs-openjdk-11` should be ready for darwin-aarch64 soon as well.
> 4. We are starting to hook up M1 machines to our build infrastructure in order to gradually produce images (starting with `labs-openjdk-17`).
> 
> [1] https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon

**Comment by lewurm** [comment]
Another update:

* https://github.com/oracle/graal/pull/4173 didn't cut it, instead @teshull fixed it properly 🥳 https://github.com/oracle/graal/commit/ac694ff2ac8056e2f830561b6a157dcd28aa483b Among some corner cases around the Linux ABI, it fixes the crash when running CTW with libgraal on macOS AArch64.
* @pminearo and I have been working on instructions to build GraalVM yourself from source. Check it out here: https://github.com/pminearo/labs_hotspot Thank you so much for your patience @pminearo!
* Adding support for runtime code generation in SVM (needed for Truffle) is currently under review: https://github.com/oracle/graal/pull/4286 I tested Espresso and GraalJS so far, looks good.
* In order to build languages however, there is still a missing dependency (`LLVM_ORG`). This is blocked right now on getting proper infrastructure in place; if you run into that and don't want to wait, ping me on Slack how to work around that temporarily 🙂

**Comment by chenzhiguo** [comment]
So long, didn't I support Apple M1?

**Comment by dr-eme** [comment]
I managed to build GraalVM thanks @pminearo instructions, thanks.
However, I'd like to know if it is possible to build a native-image component somehow, as obviously using gu tries to download this from the component catalog.

**Comment by lewurm** [comment]
> I managed to build GraalVM (the latest tag 22.0.0.2) thanks @pminearo instructions, thanks.

Thanks for testing!

> I also compiled the native-image component for this VM (within the same build environment for graalvm): `mx --env ni-ce build` don't know if the same could be achieved by adding `ni-ce` to the list of compoments in the file `graal/vm/mx.vm/libgraal` before building graalvm

This should work:
```
DYNAMIC_IMPORTS=/compiler,/sdk,/substratevm,/truffle
COMPONENTS=cmp,dis,gu,gvm,lg,ni,nic,nil,nfi-libffi,poly,polynative,sdk,svm,svml,svmnfi,tfl,tflm
EXCLUDE_COMPONENTS=libpoly
NATIVE_IMAGES=lib:jvmcicompiler,native-image,lib:native-image-agent
DISABLE_INSTALLABLES=false
```
You can check with `mx --env <envfile> graalvm-show` which things will be included in that configuration.


> However, when trying to generate a native image, I get this error:
> 
> ```
> Undefined symbols for architecture arm64:
>   "_fstat$INODE64", referenced from:
>       ___svm_vm_java_version in native-example.o

This was fixed some weeks ago: https://github.com/oracle/graal/pull/4107

Are you sure you are on the latest revision? If not, can you provide a reproducer?

**Comment by dr-eme** [comment]
Thanks for pointing out the fix and env file.
Using the latest revision it worked flawesly.
Amazing the performance obtained w.r.t running native-image through roseta.
I hope we get an official release soon.

**Comment by jperedadnr** [comment]
This [PR](https://github.com/oracle/graal/pull/4286) was just merged to support the JIT compiler on Apple Silicon, but I take it only applies to macOS_AArch64 as I don't think this would work on iOS_AAarch64?

Since this [PR](https://github.com/oracle/graal/pull/4171) was merged, `DARWIN_AARCH64` also includes now `MACOS_AARCH64` and `IOS_AARCH64`, doesn't it?

Therefore, the changes to support Apple Silicon should be applied only to the `MACOS_AARCH64` platform, or am I missing something?

**Comment by lewurm** [comment]
Just wanted to say we are still working on this 🙂   We finally got internal CI bots hooked up and we are starting to configure build lanes accordingly.  It's gonna be a close call for 22.1, let's see how it works out.  But I would say it will definitely be part of 22.2.

Also, I realised that there is a bug in Sulong around varargs handling for darwin-aarch64, so effectively it means languages other than JavaScript and Espresso (Java) will not work.  That fix most likely won't make it for 22.1, unfortunately.

**Comment by borkdude** [comment]
@lewurm Excellent! About 22.1... we can always use https://github.com/graalvm/graalvm-ce-dev-builds/releases I guess?

**Comment by lewurm** [comment]
> @lewurm Excellent! About 22.1... we can always use https://github.com/graalvm/graalvm-ce-dev-builds/releases I guess?

That's correct!  We are not quite there yet unfortunately, but I think we can still make it for 22.1 which means you should see bits popping up there at some point before the release 🙂 I'll post an update when this is the case.  We do now have builds for labsjdk, so that should ease the "build it yourself"-story a bit until then:
* JDK11: https://github.com/graalvm/labs-openjdk-11/releases/tag/jvmci-22.1-b03
* JDK17: https://github.com/graalvm/labs-openjdk-11/releases/tag/jvmci-22.1-b03

**Comment by lewurm** [comment]
I'm pleased to announce that we uploaded the first snapshot containing darwin-aarch64: https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/22.1.0-dev-20220321_2332

The following components are available for this platform:
* JVM with libgraal
* native-image
* JavaScript
* Espresso

both on JDK11 and JDK17.

Please follow the regular macOS installation steps at https://www.graalvm.org/22.0/docs/getting-started/macos, but specifically make sure you do this:
> ### On Software Notarization 
> If you are using macOS Catalina and later you may need to remove the quarantine attribute from the bits before you can use them. To do this, run the following:
> 
> ```
> $ sudo xattr -r -d com.apple.quarantine /path/to/GRAALVM_HOME
> ```

Please test and report bugs (and make sure you tag me 🙂 ).

**Comment by ppalaga** [comment]
Great news, thanks [lewurm](https://github.com/lewurm)! I wonder whether there is a way to cross-compile for darwin-aarch64 on some other platform? More specifically, I am looking for a way to compile on GitHub Actions, where darwin-aarch64 hardware is not available yet.

**Comment by borkdude** [comment]
Btw, maybe a tip: if you download GraalVM from a terminal with curl and tar, you don't have to do any xattr attribute stuff: 

```
$ curl -sLO https://github.com/graalvm/graalvm-ce-dev-builds/releases/download/22.1.0-dev-20220321_2332/graalvm-ce-java11-darwin-amd64-dev.tar.gz
$ tar xzvf graalvm-ce-java11-darwin-amd64-dev.tar.gz
```

**Comment by borkdude** [comment]
Tried out the m1 build with [babashka](https://babashka.org/) and all my tests are passing. Excellent!

```
{:test 1759, :pass 10151, :fail 0, :error 0}
```

**Comment by avocade** [comment]
> Tried out the m1 build with [babashka](https://babashka.org/) and all my tests are passing. Excellent!
> 
> ```
> {:test 1759, :pass 10151, :fail 0, :error 0}
> ```

Wow! Looking forward to supercharged `bb` very soon then! :)

**Comment by ppalaga** [comment]
> Cross-compilation is not (yet) supported (see #407). 

Thanks for pointing to #407. 

> If you want to build for M1 using GraalVM Native Image now, I guess you could set up a [self-hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners) on an M1 machine. Feel free to let me know if that's something you'd want to do so we can look into supporting `darwin-aarch64` in [the GitHub Action for GraalVM](https://github.com/marketplace/actions/github-action-for-graalvm).

Yes, thanks, that would definitely be interesting!

**Comment by bdemers** [comment]
Awesome @lewurm!

I've got a native-image build that fails on an M1 Max (this build works with 22.0.0.2.r17 through Rosetta 2) :

```txt
Error: ImageSingletons.add must not overwrite existing key com.oracle.svm.core.jdk.ProtectionDomainSupport
Existing value: com.oracle.svm.core.jdk.ProtectionDomainSupport@2e140e59
New value: com.oracle.svm.core.jdk.ProtectionDomainSupport@49a71302
com.oracle.svm.core.util.UserError$UserException: ImageSingletons.add must not overwrite existing key com.oracle.svm.core.jdk.ProtectionDomainSupport
Existing value: com.oracle.svm.core.jdk.ProtectionDomainSupport@2e140e59
New value: com.oracle.svm.core.jdk.ProtectionDomainSupport@49a71302
	at com.oracle.svm.core.util.UserError.abort(UserError.java:72)
	at com.oracle.svm.hosted.ImageSingletonsSupportImpl$HostedManagement.doAdd(ImageSingletonsSupportImpl.java:109)
	at com.oracle.svm.hosted.ImageSingletonsSupportImpl.add(ImageSingletonsSupportImpl.java:39)
	at org.graalvm.sdk/org.graalvm.nativeimage.ImageSingletons.add(ImageSingletons.java:73)
	at com.oracle.svm.hosted.ProtectionDomainFeature.afterRegistration(ProtectionDomainFeature.java:48)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$setupNativeImage$14(NativeImageGenerator.java:827)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:74)
	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:827)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:554)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:514)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:413)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:591)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:129)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:621)
```

```sh
native-image --version
GraalVM 22.1.0-dev Java 17 CE (Java Version 17.0.3+4-jvmci-22.1-b03)
```

I'm not sure how to narrow down the info you would need.
You can build this branch with r17: https://github.com/okta/okta-cli/pull/301

If you can point me in the right direction I can try to craft up a stand-alone example.

**Comment by lewurm** [comment]
@bdemers thank you, that looks like a bug indeed! I'll try to reproduce it with your okta-cli PR, but my hunch is that this is unrelated to darwin-aarch64. Could you give it a try with Rosetta again but with builds for darwin-amd64 from the same snapshot page?

Also it's probably best if you could create another github issue to discuss this further 🙂

**Comment by dr-eme** [comment]
@bdemers the issue you mention (which I'm having as well with Spring Native) is related to a new (and possibly breaking) feature of 22.1. The corresponding PR is the following
https://github.com/oracle/graal/pull/4305

With local builds (either aarch64 or amd64) before that merge, I have no issues.

**Comment by olpaw** [comment]
@dr-eme this look like you found a bug on out side. The time that `LinkAtBuildTimeSupport` would need to be available in the above stacktrace is even before `@AutomaticFeature`s get initialized. This means the `LinkAtBuildTimeFeature cannot be initialized by reular means and needs special handling. Please create a separate issue for this and let me know. Also if you can provide me with a simple reproducer that would be awesome.

**Comment by globecen** [comment]
Hello and thank you for the support of macsbooks M1.
I have an error following native compilation. It's at stage 6/7.
My error is : 
Fatal error: java.lang.NullPointerException
        at com.oracle.svm.reflect.hosted.ReflectionMetadataEncoderImpl.registerAnnotation(ReflectionMetadataEncoderImpl.java:511)
        at com.oracle.svm.reflect.hosted.ReflectionMetadataEncoderImpl.registerTypeAnnotationValues(ReflectionMetadataEncoderImpl.java:500)
        at com.oracle.svm.reflect.hosted.ReflectionMetadataEncoderImpl.addClassMetadata(ReflectionMetadataEncoderImpl.java:264)
        at com.oracle.svm.hosted.image.NativeImageCodeCache.buildRuntimeMetadata(NativeImageCodeCache.java:252)
        at com.oracle.svm.hosted.image.LIRNativeImageCodeCache.layoutMethods(LIRNativeImageCodeCache.java:106)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:640)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:514)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:413)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:591)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:129)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:621)

**Comment by fniephaus** [comment]
@globecen thanks for the report. Looks like you're running into the same issue that causes our Quarkus gates to fail at the moment (see #4388 and https://github.com/oracle/graal/runs/5654668338?check_suite_focus=true#step:8:351).

We are tracking this internally and are working on a fix. The good news is that it seems unrelated to M1 support.

**Comment by tomaskrulis** [comment]
From a newbie/basic user point of view, I have followed installation instructions and installed all components, compiled all official native images without any problem on an M1 macOS laptop.

Hopefully, it makes sense to add it here to balance little the bug reports.

Thank you and good job GraalVM devs!

**Comment by fniephaus** [comment]
> Yes, thanks, that would definitely be interesting!

@ppalaga In theory, [the GitHub Action for GraalVM](https://github.com/marketplace/actions/github-action-for-graalvm) should already work on an M1-powered self-hosted runner. All you have to do is choosing `dev` builds:
```yml
jobs:
  build:
    runs-on: self-hosted
    steps:
      - uses: actions/checkout@v2
      - uses: graalvm/setup-graalvm@v1
        with:
          version: 'dev'
          java-version: '11'
          components: 'native-image'
          github-token: ${{ secrets.GITHUB_TOKEN }}
      - name: Example step
        run: native-image --version
```
If that doesn't work, please open an issue at https://github.com/graalvm/setup-graalvm/issues/new. Thanks!

**Comment by dr-eme** [comment]
> @dr-eme this looks like you found a bug on our side. The time that `LinkAtBuildTimeSupport` would need to be available in the above stacktrace is even before `@AutomaticFeature`s get initialized. This means the `LinkAtBuildTimeFeature` cannot be initialized by regular means and needs special handling. Please create a separate issue for this and let me know. Also if you can provide me with a simple reproducer that would be awesome.

Hi @olpaw, attached the [sample.zip](https://github.com/oracle/graal/files/8332742/sample.zip) reproducer. I managed to narrow down the source of the issue. It happens when I remove one or both of these modules from my classpath `log4j-api` and/or `log4j-to-slf4j`

Run `./gradlew nativeCompile` to build
.

**Comment by mraible** [comment]
I tried this on a project that works great on macOS Intel. I discovered that I had to add an additional native hint in Spring Native for things to work on my M1. https://github.com/spring-projects-experimental/spring-native/issues/1538

**Comment by olpaw** [comment]
> > @dr-eme this looks like you found a bug on our side. The time that `LinkAtBuildTimeSupport` would need to be available in the above stacktrace is even before `@AutomaticFeature`s get initialized. This means the `LinkAtBuildTimeFeature` cannot be initialized by regular means and needs special handling. Please create a separate issue for this and let me know. Also if you can provide me with a simple reproducer that would be awesome.
> 
> Hi @olpaw, attached the [sample.zip](https://github.com/oracle/graal/files/8332742/sample.zip) reproducer. I managed to narrow down the source of the issue. It happens when I remove one or both of these modules from my classpath `log4j-api` and/or `log4j-to-slf4j`
> 
> Run `./gradlew nativeCompile` to build .

@dr-eme thanks for the repoducer. I can verify that this fails on master. I will fix this for master & the 22.1 release branch.

**Comment by vincent-dm** [comment]
I just want to add that I installed `graalvm-ce-java17-22.2.0-dev` on my M1 Macbook Pro and ran the test suites of all our projects (written in Clojure, targeting Java 17) and it runs like a charm. Fast and not a single problem.

A big thanks to everyone who contributed to this!

**Comment by usersxx** [comment]
How about releasing GraalVM with JDK 11.0.15. Based on [OpenJDK's jdk11u mainline](https://github.com/openjdk/jdk11u), JDK 11.0.15 supports [darwin aarch64]((https://github.com/openjdk/jdk11u/commit/eb0708f75aa3c196e41014addfaa5667fd940cc2)) by Implementation of [JEP 391, which is added the support of arm to MacOS](https://openjdk.java.net/jeps/391)

**Comment by usersxx** [comment]
> I just want to add that I installed `graalvm-ce-java17-22.2.0-dev` on my M1 Macbook Pro and ran the test suites of all our projects (written in Clojure, targeting Java 17) and it runs like a charm. Fast and not a single problem.
> 
> A big thanks to everyone who contributed to this!

@vincent-dm, Have you tested `graalvm-ce-java11-22.2.0-dev`? If yes, does if work fine ?

**Comment by vincent-dm** [comment]
> @vincent-dm, Have you tested graalvm-ce-java11-22.2.0-dev? If yes, does if work fine ?

I just tried it and ran some of our tests, and they work fine 👍

**Comment by usersxx** [comment]
> > @vincent-dm, Have you tested graalvm-ce-java11-22.2.0-dev? If yes, does if work fine ?
> 
> I just tried it and ran some of our tests, and they work fine 👍

Great!!

**Comment by surbhi65** [comment]
I'm getting this error when I tried to install latest version on mac m1 chip “graalvm-ce-java11-22.2.0-dev” is damaged and can’t be opened.

**Comment by wetted** [comment]
@lewurm I downloaded and installed the `22.2.0-dev-20220411_2158` version for darwin/arm64 today on my M1 Max MacBook Pro. I used it as the basis for the GraalVM part of the [Micronaut guide for Data/JDBC](https://guides.micronaut.io/latest/micronaut-data-jdbc-repository-gradle-java.html) to test, and everything seemed to work just fine as far as I can tell. I look forward to a standard install from SDKMAN. Thank you.

**Comment by auula** [comment]
This is a very cool project, I love that everyone is taking action and building a good community, graalvm has a bright future! !

**Comment by grishka** [comment]
None of the LLVM stuff seems to work for me.
```
$ export PATH=~/GraalVM/Contents/Home/bin:$PATH
$ java --version
openjdk 17.0.3 2022-04-19
OpenJDK Runtime Environment GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06)
OpenJDK 64-Bit Server VM GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06, mixed mode, sharing)
$ lli --print-toolchain-path
ERROR: com.oracle.truffle.api.CompilerDirectives$ShouldNotReachHere
org.graalvm.polyglot.PolyglotException: com.oracle.truffle.api.CompilerDirectives$ShouldNotReachHere
	at com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere(CompilerDirectives.java:570)
	at com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere(CompilerDirectives.java:490)
	at com.oracle.truffle.llvm.parser.factories.UnknownBasicPlatformCapability.getVAListType(UnknownBasicPlatformCapability.java:78)
	at com.oracle.truffle.llvm.parser.factories.BasicNodeFactory.<init>(BasicNodeFactory.java:382)
	at com.oracle.truffle.llvm.nativemode.NativeConfiguration.createNodeFactory(NativeConfiguration.java:69)
	at com.oracle.truffle.llvm.runtime.pthread.LLVMPThreadContext.lambda$new$0(LLVMPThreadContext.java:92)
	at com.oracle.truffle.llvm.runtime.LLVMLanguage.lambda$createCachedCallTarget$0(LLVMLanguage.java:714)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at com.oracle.truffle.llvm.runtime.LLVMLanguage.createCachedCallTarget(LLVMLanguage.java:714)
	at com.oracle.truffle.llvm.runtime.pthread.LLVMPThreadContext.<init>(LLVMPThreadContext.java:91)
	at com.oracle.truffle.llvm.runtime.LLVMContext.<init>(LLVMContext.java:243)
	at com.oracle.truffle.llvm.runtime.LLVMLanguage.createContext(LLVMLanguage.java:367)
	at com.oracle.truffle.llvm.runtime.LLVMLanguage.createContext(LLVMLanguage.java:90)
	at com.oracle.truffle.api.LanguageAccessor$LanguageImpl.createEnvContext(LanguageAccessor.java:263)
	at com.oracle.truffle.polyglot.PolyglotLanguageContext.ensureCreated(PolyglotLanguageContext.java:587)
	at com.oracle.truffle.polyglot.PolyglotLanguageContext.ensureInitialized(PolyglotLanguageContext.java:666)
	at com.oracle.truffle.polyglot.PolyglotContextImpl.getBindings(PolyglotContextImpl.java:1073)
	at com.oracle.truffle.polyglot.PolyglotContextDispatch.getBindings(PolyglotContextDispatch.java:98)
	at org.graalvm.polyglot.Context.getBindings(Context.java:560)
	at com.oracle.truffle.llvm.launcher.LLVMLauncher.printToolchainAPI(LLVMLauncher.java:279)
	at com.oracle.truffle.llvm.launcher.LLVMLauncher.execute(LLVMLauncher.java:252)
	at com.oracle.truffle.llvm.launcher.LLVMLauncher.launch(LLVMLauncher.java:76)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:296)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:121)
	at com.oracle.truffle.llvm.launcher.LLVMLauncher.main(LLVMLauncher.java:53)
Caused by: Attached Guest Language Frames (0)
Internal GraalVM error, please report at https://github.com/oracle/graal/issues/.
```

M1 Max, macOS 12.3.1

**Comment by lewurm** [comment]
@grishka `lli` wasn't supposed to be shipped. That was my mistake that it made it into the release, sorry about that 😞

If you feel adventurous you can try to build the WIP branch to add LLVM support: https://github.com/lewurm/graal-1/tree/burbanfo/darwin-aarch64-sulong-platform
Otherwise I hope I'll get it merged this week, and then the `-dev` builds will contain a working `lli`. I can ping this thread again once those builds are available.

**Comment by lewurm** [comment]
@grishka LLVM support for darwin-aarch64 has been merged yesterday (finaaaaally 🙂).  Can you test the latest dev build? https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/22.2.0-dev-20220518_1936

Truffleruby and Graalpython will follow soon.

**Comment by johanthoren** [comment]
I'm happy to report that I just built my first Apple Silicon native image using the dev build above.

**Comment by Cs4r** [comment]
> is it possible to build an image from graalvm arm64 for MacBook m1 in order to build native images. Note that I tried to do this to build a native image for Quarkus framework but it didn't work. I used this graalvm https://github.com/graalvm/container/tree/master/community in order to build my image, but eventually after running sam cli to test the lambda function I got this error msg="Init failed" InvokeID= error="fork/exec /var/task/bootstrap: no such file or directory"


My environment: 

**MacOS:**  12.1 Monterey 
**Chipset**: Apple M1 Pro
**Apache-maven version:** 3.6.0
**Graalvm:** graalvm-ce-java17-22.0.0.2

I built the native image following the steps here: https://github.com/quarkusio/quarkus/issues/16225#issuecomment-999549585 and when invoking the lambda locally I get the following error:

```
{“ErrorType”:“exiterror”,“errorMessage”:“RequestId:blabla error: “fork/exec /var/task/bootstrap: no such file or directory”}
```


**As a fun fact:** 

I built the native image in an Intel chipset MacBook and copied the generated **_function.zip_** in target/ in my M1 MacBook, then deployed the lambda native image locally and **_that way I could run the lambda_** seamlessly.

CC @mutasemhidmi

**Comment by auula** [comment]
Whether M1 supports perfectly, how can I get M1 chip compatible version support？Graal is a great project. Thank you, Graalvm has a bright future!

**Comment by reyesyang** [comment]
> Whether M1 supports perfectly, how can I get M1 chip compatible version support？Graal is a great project. Thank you, Graalvm has a bright future!

https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.1.0
Platform：macOS (aarch64/M1)

**Comment by rmetzger** [comment]
Are there plans to provide a `wasm-installable-svm-java17-darwin-aarch64.jar` (e.g. a wasm component for M1)?

**Comment by lewurm** [comment]
> Are there plans to provide a `wasm-installable-svm-java17-darwin-aarch64.jar` (e.g. a wasm component for M1)?

Thanks for the request, you are actually the first who asked for it 🙂 Probably 22.3

**Comment by rmetzger** [comment]
It seems that wasm on graal is still in its early days ;) 
Thanks!

**Comment by smclinden** [comment]
After installing the 22.2 version for M1 Mac running 12.4 I get the following:

“graalvm-ce-java17-22.2.0” is damaged and can’t be opened.

**Comment by cap10morgan** [comment]
> After installing the 22.2 version for M1 Mac running 12.4 I get the following:
> 
> “graalvm-ce-java17-22.2.0” is damaged and can’t be opened.

Sounds like you need to do the following (from the macOS installation instructions at https://www.graalvm.org/22.2/docs/getting-started/macos/):

Note: If you are using macOS Catalina and later you may need to remove the quarantine attribute:

 `sudo xattr -r -d com.apple.quarantine /path/to/graalvm`

**Comment by smclinden** [comment]
```
Sounds like you need to do the following (from the macOS installation instructions at https://www.graalvm.org/22.2/docs/getting-started/macos/):

Note: If you are using macOS Catalina and later you may need to remove the quarantine attribute:

sudo xattr -r -d com.apple.quarantine /path/to/graalvm
```

Tried that. It didn't fix it.

**Comment by smclinden** [comment]
So I managed to build the current master from scratch and it works, but I need to add modules like graal-js and graal-node from a source build (since there aren't precompiled jars for the Master.

It is not clear how I can add these to the VM using gu. Any assistance would be appreciated.

**Comment by cuteason** [comment]
> So I managed to build the current master from scratch and it works, but I need to add modules like graal-js and graal-node from a source build (since there aren't precompiled jars for the Master.
> 
> It is not clear how I can add these to the VM using gu. Any assistance would be appreciated.

@smclinden It seems like Node.js is not supported on macOS ARM64 yet. See [this issue](https://github.com/oracle/graaljs/issues/631).

**Comment by lewurm** [comment]
Sorry for the silence on this. See https://github.com/oracle/graaljs/issues/631 for an update. Dev builds will contain nodejs soon, and it will also be part of the next release.

**Comment by lvh** [comment]
@tmueller1178 It sounds like you have a fallback image, which is mostly unrelated to this ticket.

**Comment by lewurm** [comment]
@tmueller1178 could you please share steps to reproduce the issue you are seeing? Thanks!

**Comment by azolotko** [comment]
Hi! I noticed that native-image-llvm-backend is not yet available for darwin aarch64. Could you please put a reference to a corresponding issue to follow?

**Comment by lewurm** [comment]
> Hi! I noticed that native-image-llvm-backend is not yet available for darwin aarch64. Could you please put a reference to a corresponding issue to follow?

https://github.com/oracle/graal/issues/5491 /cc @loicottet

**Comment by jackrusher** [comment]
Using `OpenJDK Runtime Environment GraalVM CE 22.3.0 (build 19.0.1+10-jvmci-22.3-b08)`, I'm currently seeing this linker failure when trying to use `native-image` to compile some classes that use Project Panama's new FFI functionality on an M1 machine:
```
Undefined symbols for architecture arm64:
  "_Java_java_lang_invoke_MethodHandle_linkToNative", referenced from:
      ___data in ffi.o
  "_Java_jdk_internal_misc_ScopedMemoryAccess_closeScope0", referenced from:
      ___svm_version_info in ffi.o
ld: symbol(s) not found for architecture arm64
```
Would this be an expected part of the currently experimental status for M1? Should I open a separate ticket? Am I holding it wrong? :)

**Comment by lewurm** [comment]
@jackrusher my understanding is that the Foreign Linker API is generally not supported by GraalVM yet. So yes, please open a separate ticket 🙂

**Comment by bakkot** [comment]
I see on [the docs](https://www.graalvm.org/latest/docs/introduction/#features-support) that WebAssembly support on macOS ARM64 is listed as "Experimental", the same as the native image, JS, etc. But this issue says that it's only "planned", and I don't see anywhere to get it. Are the docs wrong?

**Comment by fniephaus** [comment]
@bakkot yes, it seems the docs are wrong. Thanks for pointing that out!

**Comment by kevgilmore** [comment]
Any update on native-image-llvm-backend support for M1?

**Comment by Gedochao** [comment]
Not sure if this is the right spot to ask, but - is there any reason why the `darwin-aarch64` launcher is missing from the `22.3.2` release?
https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.3.2

It seems to be there in `22.3.1`: https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.3.1

**Comment by pminearo** [comment]
I asked the same question on slack.  According to Alina Yurenko on Slack:
Unfortunately here won’t a be a 22.3.2 for macOS aarch64, since this component was experimental in 22.3. The available options are using 22.3.1 CPU, dev builds, or waiting for 23.0.

**Comment by headius** [comment]
I am testing some performance optimizations in JRuby and found that the latest 23.x development build is extremely slow to invoke a method from Ruby via Java reflection. It runs 6-7x slower than Hotspot for the following benchmark:

```ruby
Runtime = java.lang.Runtime
loop {
  t = Time.new
  i = 0
  while i < 100_000_000
    i += 1
    Runtime.runtime
  end
  puts Time.now - t
}
```

Running as follows, comparing Hotspot and GraalVM CE:

```
[] jruby $ jruby -v -Xcompile.invokedynamic -e 'Runtime = java.lang.Runtime; loop { t = Time.new; i = 0; while i < 100_000_000; i += 1; Runtime.runtime; end; puts Time.now - t }' 
jruby 9.4.3.0-SNAPSHOT (3.1.0) 2023-05-09 3fd0b8fedc OpenJDK 64-Bit Server VM 20+36 on 20+36 +indy +jit [arm64-darwin]
3.6947029999999996
3.702981
3.760052
^C%                                                                                                                                                                                                       [] jruby $ JAVA_HOME=~/graalvm-community-openjdk-20+34.1/Contents/Home jruby -v -Xcompile.invokedynamic -e 'Runtime = java.lang.Runtime; loop { t = Time.new; i = 0; while i < 100_000_000; i += 1; Runtime.runtime; end; puts Time.now - t }' 
jruby 9.4.3.0-SNAPSHOT (3.1.0) 2023-05-09 3fd0b8fedc OpenJDK 64-Bit Server VM 20+34-jvmci-23.0-b10 on 20+34-jvmci-23.0-b10 +indy +jit [arm64-darwin]
21.640044
21.707385000000002
^C%
```

You can see here 6x-7x slower performance on GraalVM CE. Is this expected given the current status of Apple M support?

**Comment by headius** [comment]
> I've opened another issue to track it: https://github.com/oracle/graal/issues/6600

Thank you!

**Comment by headius** [comment]
> I've opened another issue to track it: https://github.com/oracle/graal/issues/6600

Thank you!

**Comment by lewurm** [comment]
This is generally "fixed" 🙂  Please open new issues if you encounter any problems on this platform from now on. Thank you!
